Add incremental search to VirtualSelectBox

zaucker opened the pull request:
- Enable incremental search on the list items by `selectBox.setIncrementalSearch(true)`.
- Highlighting can be customized by setting property `highlightStyle` to a style map (defaults to red and bold).
<img width="165" alt="Screenshot 2021-01-08 at 17 32 49" src="https://user-images.githubusercontent.com/198038/104040403-e531d980-51d7-11eb-904a-607d1ee4dc0e.png">

zaucker committed:
Add incremental search

zaucker committed:
Fix comment, reenable timer

cboulanger commented:
Cool! Would you add a demo for the widget browser? 

https://github.com/qooxdoo/qxl.demobrowser/blob/master/source/class/qxl/demobrowser/demo/virtual/SelectBox.js

zaucker commented:
> Cool! Would you add a demo for the widget browser?
> 
> https://github.com/qooxdoo/qxl.demobrowser/blob/master/source/class/qxl/demobrowser/demo/virtual/SelectBox.js

Will do.

derrell commented on framework/source/class/qx/ui/form/VirtualSelectBox.js:
```diff
@@ -445,6 +467,150 @@ qx.Class.define("qx.ui.form.VirtualSelectBox",
+    __applyMarkHighlightStyle : function(value, old) {
+      var styles = [];
+      if (value != null) {
+        var keys = Object.keys(value);
```
What do you think about adding the option for highlight style to be a string that maps to an appearance in the theme, since this is a fully themeable widget?

zaucker replied:
That would make sense. Could you point me to an example?

oetiker replied:
since the highlight is created using a span in html with the 'rich' property set on the value ... this does not mesh well with qooxdoo theming

zaucker replied:
Ah. yes, we are only highlighting part of the label string of the list item.

derrell replied:
The way to pull this from the theme is as so:
```
qx.theme.manager.Appearance.getInstance().styleFrom("list-search-highlight")
```

derrell replied:
(And then we should add it to the few internal themes we ship with)

derrell replied:
As stated in the "conversation", you can do it like this:

```
qx.theme.manager.Appearance.getInstance().styleFrom("list-search-highlight")
```
and then add "list-search-highlight" to the few shipped themes.

derrell commented on framework/source/class/qx/ui/form/VirtualSelectBox.js:
```diff
@@ -445,6 +467,150 @@ qx.Class.define("qx.ui.form.VirtualSelectBox",
+              return item.match(filterValue);
+          },
+          configureItem : function(item) {
+              item.setRich(true);
```
Does this potentially allow for unanticipated injection of evil code? I understand that `rich` is a prerequisite to attaining this behavior, but wondering if we need a warning someplace that it will be automatically made `rich`? Since the search is always from the beginning of the list item value, maybe highlighting should be an option since the same early text will be highlighted in every row so is pretty but not absolutely necessary?

oetiker replied:
true, the input would have to be escaped html escaped ...

zaucker replied:
> Does this potentially allow for unanticipated injection of evil code? I understand that `rich` is a prerequisite to attaining this behavior, but wondering if we need a warning someplace that it will be automatically made `rich`? 

> Since the search is always from the beginning of the list item value, maybe highlighting should be an option since the same early text will be highlighted in every row so is pretty but not absolutely necessary? And if we don't highlight, then `rich` isn't necessary.
The search is intentionally **not** required to be from the beginning of the string. My use case is a select box containing addresses as items and always searching from the beginning is not very useful, especially of the item starts with a zip code or a town, because at least in our customer's application the street is much more likely to be "relatively" unique.
And thus the highlighting is not just pretty but necessary for the user to recognize why an item matches.
<img width="143" alt="Screenshot 2021-01-08 at 20 27 25" src="https://user-images.githubusercontent.com/198038/104056041-1918f900-51f0-11eb-858b-cf38a1c66ad1.png">

zaucker replied:
> true, the input would have to be escaped html escaped ...

Well, the only HTML that could be entered is what's already there in the item labels. Is this really an issue?

zaucker replied:
But I added `escapeHtml()` around the filter value.

oetiker replied:
the problem is that the list labels are normally not 'rich' ... escapeHtml should take care of the problem

zaucker replied:
> the problem is that the list labels are normally not 'rich' ... escapeHtml should take care of the problem

That would break the use of HTML formatted item labels together with incremental search.

zaucker replied:
Fixed, I think.

derrell reviewed (changes requested):
Very cool feature! I just added a potential security issue question that should probably be considered before merging.

zaucker committed:
Rename method

zaucker committed:
HTML escape search string

derrell commented on framework/source/class/qx/ui/form/VirtualSelectBox.js:
```diff
@@ -445,6 +467,151 @@ qx.Class.define("qx.ui.form.VirtualSelectBox",
+
+    _highlightFilterValue : function (item)
+    {
+      return this.__highlightStyle ? item.replace(item, '<span style="' + this.__highlightStyle + '">' + qx.module.util.String.escapeHtml(this.__filterValue) + '</span>')
```
This is good. It solves the problem... except in the case where the user had turned on `rich` intentionally, in which case we shouldn't escape it. So something like this, maybe:

```
  return (
    this.__highlightStyle
      ? item.replace(
          item,
          [
            '<span style="',
            this.__highlightStyle,
            '">',
            this.__lastRich
              ? qx.module.util.String.escapeHtml(this.__filterValue)
              : this.__filterValue,
            '</span>'
          ].join(""))
      : item);
```

zaucker replied:
Done.

zaucker replied:
Get highlight style via theming appearance and added the style to all the themes.

oetiker replied:
hmm if there is 'evil' html outside the highlighted part, it will not get escaped I think

zaucker replied:
> hmm if there is 'evil' html outside the highlighted part, it will not get escaped I think

Yes, as it wouldn't if the programmer has set them to rich himself (which is done exactly like here in one of the demobrowser examples) and doesn't take care what content he has in the model.

zaucker committed:
Add highlight style appearance and fix escaping

zaucker committed:
Fix highlight appearances

zaucker committed:
Merge branch 'master' into VirtualSelectBoxIncrementalSearch

cboulanger commented on framework/source/class/qx/ui/form/VirtualSelectBox.js:
```diff
@@ -519,26 +517,40 @@ qx.Class.define("qx.ui.form.VirtualSelectBox",
+    __getHighlightStyle : function() {
+      var highlightAppearance =
+        qx.theme.manager.Appearance.getInstance().styleFrom("list-search-highlight");
+      console.log('__getHighlightStyle=', highlightAppearance);
```
Is this still there? Should be removed before merging.

zaucker replied:
Thanks for the catch. Removed.

derrell reviewed (approved):
Very cool feature! Thanks!

zaucker committed:
Remove debug output

zaucker committed:
Merge branch 'VirtualSelectBoxIncrementalSearch' of github.com:zaucker/qooxdoo into VirtualSelectBoxIncrementalSearch

oetiker commented on framework/source/class/qx/ui/form/VirtualSelectBox.js:
```diff
@@ -70,6 +74,22 @@ qx.Class.define("qx.ui.form.VirtualSelectBox",
+      check : "Boolean"
+    },
+
+    highlightStyle :
+    {
+        nullable : true,
+        apply : '__applyHighlightStyle',
+        init : {
+            color      : '#FF0000',
+            fontWeight : 'bold'
+        }
+    },
```
So this does not seem to get used anymore ?

zaucker replied:
Indeed, thanks. Removed.

zaucker committed:
Remove unused highlightStyle property

zaucker committed:
Syntax fix

zaucker commented:
> > Cool! Would you add a demo for the widget browser?
> > https://github.com/qooxdoo/qxl.demobrowser/blob/master/source/class/qxl/demobrowser/demo/virtual/SelectBox.js
> 
> Will do.

https://github.com/qooxdoo/qxl.demobrowser/pull/12

zaucker committed:
- Use hyphenate() instead of regex to convert from camelCase
- htmlEscape() whole item label if rich wasn't set before
- Document rich behaviour in with property incrementalSearch
- Fix some indentations
Fix indentation.

zaucker committed:
- Use hyphenate() instead of regex to convert from camelCase
- htmlEscape() whole item label if rich wasn't set before
- Document rich behaviour in with property incrementalSearch
- Fix some indentations

zaucker committed:
Merge branch 'VirtualSelectBoxIncrementalSearch' of github.com:zaucker/qooxdoo into VirtualSelectBoxIncrementalSearch

zaucker committed:
Make incrementalSearch apply method protected

zaucker commented:
> Shouldn't an `apply` method be protected rather than private? Since an derived class might want to modify it...

Done.

hkollmann committed:
Merge branch 'master' into VirtualSelectBoxIncrementalSearch

hkollmann commented:
some open linting issues:

Warning: qx.ui.form.VirtualSelectBox (594, 7): Using an undeclared private class feature: '__highlightStyle'
  - Warning: qx.ui.form.VirtualSelectBox (535, 28): Using an undeclared private class feature: '__highlightStyle'
  - Warning: qx.ui.form.VirtualSelectBox (500, 49): Loop or condition statement without a block as body
  - Warning: qx.ui.form.VirtualSelectBox (536, 29): Loop or condition statement without a block as body

zaucker committed:
Fix lint issues

zaucker committed:
Merge branch 'VirtualSelectBoxIncrementalSearch' of github.com:zaucker/qooxdoo into VirtualSelectBoxIncrementalSearch

zaucker commented:
> some open linting issues:
> 
> Warning: qx.ui.form.VirtualSelectBox (594, 7): Using an undeclared private class feature: '__highlightStyle'
> 
> * Warning: qx.ui.form.VirtualSelectBox (535, 28): Using an undeclared private class feature: '__highlightStyle'
> * Warning: qx.ui.form.VirtualSelectBox (500, 49): Loop or condition statement without a block as body
> * Warning: qx.ui.form.VirtualSelectBox (536, 29): Loop or condition statement without a block as body

Fixed.

level420 reviewed (approved):
Nice feature! Thank you @zaucker !

zaucker commented:
> Nice feature! Thank you @zaucker !

Thanks!

zaucker commented:
Just for documentation the latest theming:
<img width="145" alt="Screenshot 2021-01-11 at 15 28 24" src="https://user-images.githubusercontent.com/198038/104195767-51ece400-5423-11eb-8149-9706d36367e2.png">
